cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -414,30 +520,46 @@ public boolean apply(RunRecordMeta input) {
   }
 
-  private Map<ProgramRunId, RunRecordMeta> getActiveRuns(ProgramId programId, final long startTime, final long endTime,
-                                                          int limit, @Nullable Predicate<RunRecordMeta> filter) {
+  private Map<ProgramRunId, RunRecordMeta> getActiveRuns(@Nullable ProgramId programId, final long startTime,
+                                                         final long endTime, int limit,
+                                                         @Nullable Predicate<RunRecordMeta> filter) {
     return getNonCompleteRuns(programId, TYPE_RUN_RECORD_STARTED, startTime, endTime, limit, filter);
   }
 
-  private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(ProgramId programId, String recordType,
-                                                 final long startTime, final long endTime, int limit,
-                                                 Predicate<RunRecordMeta> filter) {
-    MDSKey activeKey = getProgramKeyBuilder(recordType, programId).build();
+  private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(@Nullable ProgramId programId, String recordType,
+                                                              final long startTime, final long endTime, int limit,
+                                                              Predicate<RunRecordMeta> filter) {
+    Predicate<RunRecordMeta> valuePredicate = andPredicate(new Predicate<RunRecordMeta>() {
 
-    Map<MDSKey, RunRecordMeta> runRecordMap = listKV(activeKey, null, RunRecordMeta.class, limit,
-                                                     andPredicate(new Predicate<RunRecordMeta>() {
       @Override
       public boolean apply(RunRecordMeta input) {
         return input.getStartTs() >= startTime && input.getStartTs() < endTime;
       }
-    }, filter));
+    }, filter);
 
-    return getProgramRunIdMap(runRecordMap);
+    if (programId == null || !programId.getVersion().equals(ApplicationId.DEFAULT_VERSION)) {
+      MDSKey key = getProgramKeyBuilder(recordType, programId).build();
+      return getProgramRunIdMap(listKV(key, null, RunRecordMeta.class, limit, valuePredicate));
+    }
+
+    Predicate<MDSKey> keyPredicate = new AppVersionPredicate(ApplicationId.DEFAULT_VERSION);
+    MDSKey key = getVersionLessProgramKeyBuilder(recordType, programId).build();
+    Map<MDSKey, RunRecordMeta> oldRecords = listKV(key, null, RunRecordMeta.class, limit, keyPredicate, valuePredicate);
+
+    int remaining = limit - oldRecords.size();
+    if (remaining > 0) {
+      // We need to scan twice since the scan key is modified based on whether we include the app version or not.
+      key = getProgramKeyBuilder(recordType, programId).build();
+      Map<MDSKey, RunRecordMeta> newRecords = listKV(key, null, RunRecordMeta.class, remaining, keyPredicate,
+                                                     valuePredicate);
+      oldRecords.putAll(newRecords);
```

Comment:
Didn't find any guarantees, but I will switch the scan orders to make sure the order is from new -> old